doc/configuration.md
# Configuration Manual

中文配置手册请参考[这里](configuration_cn.md)  

**nexus** uses GFlags library to parse and hold command line parameters. For intergrated SDK, GFlags function `google::SetCommandLineOption("option", "value")` can be used to set options without using GFlags to parse command line.

Following are the configuration iterms in cluster node and client side.

## Cluster

| Options                        | Default    | Explanation                                                     |
| ------------------------------ | ---------- | --------------------------------------------------------------- |
| `cluster_members`              | `""`       | comma divided member list. e.g.: 127.0.0.1:8868,127.0.0.1:8869  |
| `server_id`                    | `1`        | offset of current node in member list                           |
| `ins_data_dir`                 | `"data"`   | path to store data                                              |
| `ins_binlog_dir`               | `"binlog"` | path to store raft log                                          |
| `max_cluster_size`             | `10`       | max size of cluster, must be bigger than member list size       |
| `log_rep_batch_max`            | `500`      | max number of raft log in a single log replication request      |
| `replication_retry_timespan`   | `2000`     | wait time before retrying a failed replication in ms            |
| `elect_timeout_min`            | `150`      | min time of an election timeout in ms                           |
| `elect_timeout_max`            | `300`      | max time of an election timeout in ms                           |
| `session_expire_timeout`       | `6000000`  | time to decide a session timeout in us                          |
| `max_write_pending`            | `10000`    | max size of write queue, overflow will lead to write denial     |
| `max_commit_pending`           | `10000`    | max size of commit queue, overflow will lead to request denial  |
| `ins_data_compress`            | `true`     | whether data will be compressed before written to leveldb       |
| `ins_binlog_compress`          | `true`     | whether raft log will be compressed before written to leveldb   |
| `ins_gc_interval`              | `60`       | interval of garbage collection task in second                   |
| `ins_max_throughput_in`        | `-1`       | max in throughput of one node in MB/s, -1 represents unlimited  |
| `ins_max_throughput_out`       | `-1`       | max out throughput of one node in MB/s, -1 represents unlimited |
| `ins_data_block_size`          | `4`        | block size of data leveldb in MB                                |
| `ins_binlog_block_size`        | `4`        | block size of raft log leveldb in MB                            |
| `ins_data_write_buffer_size`   | `4`        | write buffer size of data leveldb in MB                         |
| `ins_binlog_write_buffer_size` | `4`        | write buffer size of raft log leveldb in MB                     |
| `performance_interval`         | `1000`     | interval of rpc statistic updating in ms                        |
| `performance_buffer_size`      | `60`       | buffer size of rpc statistics                                   |
| `ins_trace_ratio`              | `0.001`    | ratio of sampling rpc calling log                               |
| `ins_log_file`                 | `stdout`   | path to save log, stdout represents using stdout                |
| `ins_log_size`                 | `1024`     | single file size limit of log in MB                             |
| `ins_log_total_size`           | `10240`    | overall file size limit of log in MB                            |

## SDK

SDK can use same flag file as node side, and options except following will be ignored. Following options can also be in node side flag file and will alse be ignored.

| Options                    | Default   | Explanation                                                                                 |
| -------------------------- | --------- | ------------------------------------------------------------------------------------------- |
| `cluster_members`          | `""`      | comma divided member list. e.g.: 127.0.0.1:8868,127.0.0.1:8869                              |
| `ins_log_file`             | `stdout`  | path to save log, stdout represents using stdout                                            |
| `ins_log_size`             | `1024`    | single file size limit of log in MB                                                         |
| `ins_log_total_size`       | `10240`   | overall file size limit of log in MB                                                        |
| `ins_watch_timeout`        | `120`     | interval of watch rpc in second                                                             |
| `ins_backup_watch_timeout` | `115`     | interval of backup watch in second                                                          |
| `ins_sdk_session_timeout`  | `6000000` | time to decide a session timeout in us, should not be bigger than `session_expired_timeout` |

## Client(Old Version)

Following options are used only in old version of client. The wrapper scripts are designed to shield all this options. And options in SDK is also useful.

| Options                    | Default   | Explanation                                                        |
| -------------------------- | --------- | ------------------------------------------------------------------ |
| `ins_cmd`                  | `""`      | command for client                                                 |
| `ins_key`                  | `"key"`   | key for get/put/lock/unlock/watch/register/login/logout operations |
| `ins_value`                | `"v123"`  | value for put/register/login/logout operations                     |
| `ins_start_key`            | `""`      | start key for scan operation                                       |
| `ins_end_key`              | `""`      | end key for scan operation                                         |
| `ins_rm_binlog_server_id`  | `""`      | server id to clean raft log(Deprecated)                            |
| `ins_rm_binlog_index`      | `0`       | index to clean raft log(Deprecated)                                |

## Client(New Version)

New designed client uses following options, and will not share these options with SDK and node. Also the client uses options in SDK.

| Options   | Default  | Explanation                                                         |
| --------- | -------- | ------------------------------------------------------------------- |
| `v`       | `false`  | enable verbose mode to output logs in all log level                 |
| `i`       | `false`  | enable interactive mode, data output format will be sightly changed |
| `h`       | `false`  | show help info                                                      |



doc/configuration_cn.md
# 配置信息

For English version of this manual, please refer to [this](configuration.md)

**nexus**使用GFlags来处理命令行参数。对于SDK的使用者，如果不希望使用GFlags来处理命令行参数，可以通过使用`google::SetCommandLineOption("option", "value")`来修改下面提到的配置项。

下面列出集群和客户端使用的配置项。

## 集群配置

| 配置项                         | 默认值     | 作用                                                    |
| ------------------------------ | ---------- | ------------------------------------------------------- |
| `cluster_members`              | `""`       | 逗号分隔的集群成员，如example.com:8000,example.com:9000 |
| `server_id`                    | `1`        | 当前节点在集群成员中的偏移量                            |
| `ins_data_dir`                 | `"data"`   | 数据存放路径                                            |
| `ins_binlog_dir`               | `"binlog"` | 同步的log存放路径                                       |
| `max_cluster_size`             | `10`       | nexus集群最大节点数量                                   |
| `log_rep_batch_max`            | `500`      | 批量日志同步时单次同步最大值                            |
| `replication_retry_timespan`   | `2000`     | 日志同步失败后重试等待时间                              |
| `elect_timeout_min`            | `150`      | 选举超时时间的最小值，单位ms                            |
| `elect_timeout_max`            | `300`      | 选举超时时间的最大值，单位ms                            |
| `session_expire_timeout`       | `6000000`  | 客户端session超时时间，单位us                           |
| `max_write_pending`            | `10000`    | 写操作队列最大长度，超出会拒绝写请求                    |
| `max_commit_pending`           | `10000`    | commit队列最大长度，超出会拒绝日志同步                  |
| `ins_data_compress`            | `true`     | 数据写入leveldb时是否压缩                               |
| `ins_binlog_compress`          | `true`     | 同步的log写入leveldb时是否压缩                          |
| `ins_gc_interval`              | `60`       | 垃圾清零时间间隔，单位s                                 |
| `ins_max_throughput_in`        | `-1`       | nexus集群最大入带宽限制，单位MB/s，-1表示无限制         |
| `ins_max_throughput_out`       | `-1`       | nexus集群最大出带宽限制，单位MB/s，-1表示无限制         |
| `ins_data_block_size`          | `4`        | 数据存储leveldb块大小，单位MB                           |
| `ins_binlog_block_size`        | `4`        | 同步的log存储leveldb块大小，单位MB                      |
| `ins_data_write_buffer_size`   | `4`        | 数据存储leveldb写缓冲区大小，单位MB                     |
| `ins_binlog_write_buffer_size` | `4`        | 同步的log存储leveldb写缓冲区大小，单位MB                |
| `performance_interval`         | `1000`     | rpc数据统计单位时间，单位ms                             |
| `performance_buffer_size`      | `60`       | rpc数据统计缓冲区大小                                   |
| `ins_trace_ratio`              | `0.001`    | rpc调用时输出调用者地址到日志到概率                     |
| `ins_log_file`                 | `stdout`   | 日志文件位置，如为stdout则输出到标准输出                |
| `ins_log_size`                 | `1024`     | 日志文件单个文件大小限制，单位MB                        |
| `ins_log_total_size`           | `10240`    | 日志文件总大小限制，单位MB                              |

## SDK配置

SDK可以使用和集群相同的配置文件，多余的选项则会被忽略。集群端也可以设置下面的选项，同样也会被忽略。

| 配置项                     | 默认值    | 作用                                                                 |
| -------------------------- | --------- | -------------------------------------------------------------------- |
| `cluster_members`          | `""`      | 逗号分隔的集群成员，如example.com:8000,example.com:9000              |
| `ins_log_file`             | `stdout`  | 日志文件位置，如为stdout则输出到标准输出                             |
| `ins_log_size`             | `1024`    | 日志文件单个文件大小限制，单位MB                                     |
| `ins_log_total_size`       | `10240`   | 日志文件总大小限制，单位MB                                           |
| `ins_watch_timeout`        | `120`     | watch操作单次rpc超时时间，单位s                                      |
| `ins_backup_watch_timeout` | `115`     | 备份watch操作单词rpc超时时间，单位s                                  |
| `ins_sdk_session_timeout`  | `6000000` | sdk与集群的session超时时间，不应大于`session_expire_timeout`，单位us |

## 客户端（旧版）

下面的参数在旧版客户端中生效，为方便用户使用，使用脚本来封装了这些参数。

| 配置项                     | 默认值    | 作用                                            |
| -------------------------- | --------- | ----------------------------------------------- |
| `ins_cmd`                  | `""`      | 指定客户端的操作                                |
| `ins_key`                  | `"key"`   | 指定key用于get/lock/unlock/login/logout等操作   |
| `ins_value`                | `"v123"`  | 指定value用于get/lock/unlock/login/logout等操作 |
| `ins_start_key`            | `""`      | 指定start key用于scan操作                       |
| `ins_end_key`              | `""`      | 指定end key用于scan操作                         |
| `ins_rm_binlog_server_id`  | `""`      | 指定server id用于清理同步的log                  |
| `ins_rm_binlog_index`      | `0`       | 清理指定index前的所有同步的log                  |

## 客户端（新版）

新版的客户端使用的参数不与集群和SDK公用。除下面参数外还会使用SDK的参数。

| 配置项   | 默认值   | 作用                                                   |
| -------- | -------- | ------------------------------------------------------ |
| `v`      | `false`  | 设置后会输出所有级别的log，默认值输出warning及以上级别 |
| `i`      | `false`  | 交互模式，打开后部分操作的输出格式会有所不同           |
| `h`      | `false`  | 输出帮助信息                                           |



src/server/flags.cc
#include <gflags/gflags.h>

DEFINE_string(cluster_members, "", "cluster members , e.g. abc.com:1234,def.com:3456");
DEFINE_int32(server_id, 1, "the offset in cluster members of this node");
DEFINE_string(ins_data_dir, "data", "local directory which store pesistent information");
DEFINE_string(ins_binlog_dir, "binlog", "write-ahead log directory path");
DEFINE_int32(max_cluster_size, 10, "maximum size of ins cluster");
DEFINE_int32(log_rep_batch_max, 500, "maximum batch size of log replication");
DEFINE_int32(replication_retry_timespan, 2000, "when replication fail, sleep a while before retry");
DEFINE_int64(elect_timeout_min, 150, "mininum timeout to make a new election");
DEFINE_int32(elect_timeout_max, 300, "maximum timeout to make a new election");
DEFINE_int64(session_expire_timeout, 6000000, "timeout for session expiration, 6 seconds in default");
DEFINE_int32(max_write_pending, 10000, "max write pending size of Put");
DEFINE_int32(max_commit_pending, 10000, "max commit pending size");
DEFINE_bool(ins_data_compress, true, "enable snappy compression on leveldb storage");
DEFINE_bool(ins_binlog_compress, true, "enable snappy compression on binlog");
DEFINE_int32(ins_gc_interval, 60, "binlog clean interval (seconds)");
DEFINE_int32(ins_max_throughput_in, -1, "max input throughput, MB");
DEFINE_int32(ins_max_throughput_out, -1, "max output throughput, MB");
DEFINE_int32(ins_data_block_size, 4, "for data, leveldb block_size, KB");
DEFINE_int32(ins_binlog_block_size, 4, "for binlog, leveldb block_size, KB");
DEFINE_int32(ins_data_write_buffer_size, 4, "for data, leveldb write_buffer_size, MB");
DEFINE_int32(ins_binlog_write_buffer_size, 4, "for binlog, leveldb write_buffer_size, MB");
DEFINE_int32(performance_interval, 1000, "milliseconds of the interval of performance counter ticktock");
DEFINE_int32(performance_buffer_size, 60, "size of the buffer to hold the history record of performance data");
DEFINE_double(ins_trace_ratio, 0.001, "trace log printing ratio");
DEFINE_string(ins_log_file, "stdout", "filename of log file");
DEFINE_int32(ins_log_size, 1024, "max size of single log file");
DEFINE_int32(ins_log_total_size, 10240, "max size of all log file");

//ins_cli only
DEFINE_string(ins_cmd, "", "the command of inc shell");
DEFINE_string(ins_key, "key", "");
DEFINE_string(ins_start_key, "", "start key of scan");
DEFINE_string(ins_end_key, "", "end key of scan (excluded)");
DEFINE_string(ins_value, "v123", "");
DEFINE_int64(ins_rm_binlog_index, 0, "end index of binlog clean operation");
DEFINE_string(ins_rm_binlog_server_id, "", "servier id of binlog clean operation");
DEFINE_int32(ins_watch_timeout, 120, "wath timeout(seconds)");
DEFINE_int32(ins_backup_watch_timeout, 115, "backup watch timeout(seconds)");
DEFINE_int64(ins_sdk_session_timeout, 6000000, "timeout for session expiration in sdk side");
